skimage/future/_thin_plate_splines.py

```diff
@@ -0,0 +1,323 @@
+import numpy as np
+import scipy as sp
+
+from .._shared.utils import check_nD
+
+
+class TpsTransform:
+    """Thin-plate spline transformation.
+
+    Given a set of control points (source and destination points), this class
+    can be used to estimate the thin-plate spline (TPS) transformation, which
+    transforms the source points into the destination points.
+    It interpolates a surface that passes through each control point.
+    Control points are seen as position constraints on a bending surface. The ideal
+    surface is the one that bends the least.
+
+    Attributes
+    ----------
+    spline_mappings : (N, D) array_like
+        Spline mappings for every control point. These describe a deformation.
+    src : (N, 2) array_like
+        Coordinates of control points in source image.
+
+    References
+    ----------
+    .. [1] Bookstein, Fred L. "Principal warps: Thin-plate splines and the
+           decomposition of deformations," IEEE Transactions on pattern analysis
+           and machine intelligence 11.6 (1989): 567–585.
+           DOI:`10.1109/34.24792`
+           https://user.engineering.uiowa.edu/~aip/papers/bookstein-89.pdf
+
+    Examples
+    --------
+    >>> import skimage as ski
+
+    Define source and destination control points:
+
+    >>> src = np.array([[0, 0], [0, 5], [5, 5],[5, 0]])
+    >>> dst = np.roll(src, 1, axis=0)
+
+    Generate meshgrid:
+
+    >>> coords = np.meshgrid(np.arange(5), np.arange(5))
+    >>> t_coords = np.vstack([coords[0].ravel(), coords[1].ravel()]).T
+
+    Estimate transformation:
+
+    >>> tps = ski.future.TpsTransform()
+    >>> tps.estimate(src, dst)
+    True
+
+    Apply the transformation:
+
+    >>> trans_coord = tps(t_coords)
+    >>> xx_trans = trans_coord[:, 0]
+    >>> yy_trans = trans_coord[:, 1]
+    >>> coords[1]
+    array([[0, 0, 0, 0, 0],
+           [1, 1, 1, 1, 1],
+           [2, 2, 2, 2, 2],
+           [3, 3, 3, 3, 3],
+           [4, 4, 4, 4, 4]])
+    >>> expected_yy = np.array([0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4,
+    ...                         0, 1, 2, 3, 4])
+    >>> np.allclose(yy_trans, expected_yy)
+    True
+    """
+
+    def __init__(self):
+        self._estimated = False
+        self.spline_mappings = None
```

Comment:
Unless these are generically useful outside of the transform, hide it for now as a private member?